### ISSUE TITLE
String `.endsWith` & `.startsWith` could perhaps be typed more strongly

### ISSUE THREAD

**Issue by alii** [colinhacks/zod#1747]
If the passed value was a literal, we could have a generic in ZodString that will include the prefix.

Use case:
At the moment, I am using the code below to validate that a value is a specific ID format that has a constant prefix

```ts
export function id<P extends IdPrefixes>(prefix: P) {
	return z
		.string()
		.refine((value): value is `${P}_${string}` => pika.validate(value, prefix));
}
```

Imagine if this could be done with just `.startsWith` and `.endsWith` — ZodString could then have a single generic of the output, which would default to just a string and can be made more specific with these two methods.

<img width="785" alt="CleanShot 2022-12-23 at 14 02 47@2x" src="https://user-images.githubusercontent.com/25351731/209348315-61d308e4-6581-4210-91ed-b02398adb85c.png">

**Comment by santosmarco-caribou** [comment]
I've opened a PR for that. Let's see what people think.

https://github.com/colinhacks/zod/pull/1751

**Comment by colinhacks** [comment]
I have some reservations discussed here: Thanks for the PR. I have some reservations discussed here: https://github.com/colinhacks/zod/pull/1730#issuecomment-1364490384

**Comment by JacobWeisenburger** [comment]
I really like the idea of better static types for `.startsWith` and `.startsWith`. But I think @colinhacks is right, it would be quite hard to do this in a backwards compatible way.

@alii or @santosmarco-caribou, do you have some ideas on how to address backwards compatibility?

**Comment by santosmarco-caribou** [comment]
One way would be to add an overload to these methods (almost like I did on my PR for `ZodRequired`).

We could leave the current implementation at the top of the overloads so the users will not feel any difference. This type of "exact" inference would be turned off by default.

Then, as a second overload, we add a definition that forces the user to pass `true` in one of the properties of the `options` object, turning the inference on.

Something like this:

```ts
startsWith(prefix: string, options?: { message?: string; exact?: never }): ZodString
startsWith<Start extends string>(prefix: Start, options: { message?: string; exact: true }): ZodString<Start, End>
startsWith(prefix: string, options?: { message?: string; exact?: true }): ZodString {
  // no change in implementation, the `exact` is solely for turning the inference on.
}
```

**Comment by JacobWeisenburger** [comment]
I like it.

Any objections/reservations @colinhacks?

**Comment by tachibanayui** [comment]
> One way would be to add an overload to these methods (almost like I did on my PR for `ZodRequired`).
> 
> We could leave the current implementation at the top of the overloads so the users will not feel any difference. This type of "exact" inference would be turned off by default.
> 
> Then, as a second overload, we add a definition that forces the user to pass `true` in one of the properties of the `options` object, turning the inference on.
> 
> Something like this:
> 
> ```ts
> startsWith(prefix: string, options?: { message?: string; exact?: never }): ZodString
> startsWith<Start extends string>(prefix: Start, options: { message?: string; exact: true }): ZodString<Start, End>
> startsWith(prefix: string, options?: { message?: string; exact?: true }): ZodString {
>   // no change in implementation, the `exact` is solely for turning the inference on.
> }
> ```

I think this is a breaking change because the second argument was `message?: string`.

**Comment by tachibanayui** [comment]
> One way would be to add an overload to these methods (almost like I did on my PR for `ZodRequired`).
> 
> We could leave the current implementation at the top of the overloads so the users will not feel any difference. This type of "exact" inference would be turned off by default.
> 
> Then, as a second overload, we add a definition that forces the user to pass `true` in one of the properties of the `options` object, turning the inference on.
> 
> Something like this:
> 
> ```ts
> startsWith(prefix: string, options?: { message?: string; exact?: never }): ZodString
> startsWith<Start extends string>(prefix: Start, options: { message?: string; exact: true }): ZodString<Start, End>
> startsWith(prefix: string, options?: { message?: string; exact?: true }): ZodString {
>   // no change in implementation, the `exact` is solely for turning the inference on.
> }
> ```

I think this is a breaking change because the second argument is `message?: string | {message?: string }`. It should be:
```ts
startsWith(prefix: string, options?: string | { message?: string; exact?: never }): ZodString
startsWith<Start extends string>(prefix: Start, options: { message?: string; exact: true }): ZodString<Start, End>
startsWith(prefix: string, options?: string | { message?: string; exact?: true }): ZodString {
  // no change in implementation, the `exact` is solely for turning the inference on.
}
```
IMO, I prefer creating a separate method for template literals because it looks cleaner when chaining methods:
```ts
const schemaUrl = z
  .string()
  .startsWith("https://", {message: "Protocol must be https!", exact: true})
  .endsWith("__schema.json", {exact: true})
  .max(255);

const schemaUrl = z
  .string()
  .startsWithLiteral("https://", "Protocol must be https!")
  .endsWithLiteral("__schema.json")
  .max(255);

**Comment by alii** [comment]
I just had a thought, perhaps the existence of `z.ExactString` could resolve this...

For example, ZodString could be the default for all `z.string()` and then when calling `.endsWith` & `.startsWith` we return a `z.ExactString<Start, Mid, End>` as documented in #1730

**Comment by santosmarco-caribou** [comment]
Yeah, I like the idea of a `ZodTemplate` (as in "template string") as well.

You pass an array of components (mainly `ZodLiteral`s but we can extend it to receive other types as well) to it (or even raw values), and the inference results in a template string.

![image](https://user-images.githubusercontent.com/98408205/209486326-3a6fb78a-6a9d-4951-9e2e-3378015c6c6c.jpeg)

But that's a bit complex to handle.

**Comment by tachibanayui** [comment]
I made a type that returns `ZodExactString<"","","">`  or `ZodString` like @alii said without creating 2 classes [here](https://github.com/colinhacks/zod/pull/1730#issuecomment-1364543411)
